core/services/bulletprooftxmanager/eth_confirmer.go

```diff
@@ -1208,3 +1222,31 @@ func findEthTxWithNonce(db *gorm.DB, fromAddress gethCommon.Address, nonce uint)
 	}
 	return &etx, errors.Wrap(err, "findEthTxsWithNonce failed")
 }
+
+func (ec *EthConfirmer) ResumePendingTaskRuns(ctx context.Context, head models.Head) error {
+	sqlxDB := postgres.UnwrapGormDB(ec.db)
+
+	type x struct {
+		ID      uuid.UUID
+		Receipt []byte
+	}
+	var receipts []x
+	if err := sqlxDB.Select(&receipts, `
+	SELECT pipeline_task_runs.id, eth_receipts.receipt FROM pipeline_task_runs
+	INNER JOIN pipeline_runs ON pipeline_runs.id = pipeline_task_runs.pipeline_run_id
+	INNER JOIN eth_txes ON (eth_txes.meta->>'PipelineTaskRunID')::uuid = pipeline_task_runs.id AND eth_txes.state = 'confirmed'
+	INNER JOIN eth_tx_attempts ON eth_txes.id = eth_tx_attempts.eth_tx_id
+	INNER JOIN eth_receipts ON eth_tx_attempts.hash = eth_receipts.tx_hash
+	WHERE pipeline_runs.state = 'suspended' AND eth_receipts.block_number <= ($1 - (eth_txes.meta->>'MinConfirmations')::integer)
+	`, head.Number); err != nil {
+		return err
+	}
+
+	for _, data := range receipts {
```

Comment:
I'm a bit worried about the scalability of this - can we add a benchmark test? For some context, all these txes correspond to vrf pipeline runs https://polygonscan.com/txs?a=0x3d2341ADb2D31f1c5530cDC622016af293177AE0. So I'm curious how slow it runs with perhaps 100k pipeline runs (the default pipeline reaping interval is 24h), 1M eth txes (afaik we never reap ethtxes) and 1s (polygon block interval) block times

Reply:
I'm a bit concerned too but it's not too different than what was executed per pending eth_tx before. @samsondav said we should be able to optimize the queries enough that it should be fine. I think with proper indexing we should be able to quickly narrow down the set of txes to scan since we can eliminate any without a task run ID

Reply:
Let's make a mental note to go over this together, carefully, and run some test queries on a large database before this gets merged.

Reply:
Here's a test using a data dump (30k pipeline runs, 700 eth_txes, I set 4 runs to suspended and made some links to eth_txes):

```
                                                                                          QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=1.24..25.69 rows=1 width=1216) (actual time=0.035..0.047 rows=1 loops=1)
   Join Filter: (eth_receipts.block_number <= (30000000 - eth_txes.min_confirmations))
   ->  Nested Loop  (cost=0.97..24.79 rows=1 width=53) (actual time=0.027..0.038 rows=1 loops=1)
         ->  Nested Loop  (cost=0.69..23.43 rows=1 width=28) (actual time=0.022..0.032 rows=1 loops=1)
               ->  Nested Loop  (cost=0.42..21.87 rows=1 width=16) (actual time=0.011..0.019 rows=4 loops=1)
                     ->  Index Only Scan using pipeline_runs_suspended on pipeline_runs  (cost=0.14..8.16 rows=1 width=8) (actual time=0.004..0.007 rows=4 loops=1)
                           Heap Fetches: 4
                     ->  Index Scan using pipeline_task_runs_pipeline_run_id_dot_id_idx on pipeline_task_runs  (cost=0.27..13.67 rows=4 width=24) (actual time=0.001..0.002 rows=1 loops=4)
                           Index Cond: (pipeline_run_id = pipeline_runs.id)
               ->  Index Scan using eth_txes_pipeline_task_run_id_key on eth_txes  (cost=0.28..1.56 rows=1 width=28) (actual time=0.002..0.002 rows=0 loops=4)
                     Index Cond: (pipeline_task_run_id = pipeline_task_runs.id)
         ->  Index Scan using idx_eth_tx_attempts_unique_gas_prices on eth_tx_attempts  (cost=0.28..1.35 rows=1 width=41) (actual time=0.004..0.004 rows=1 loops=1)
               Index Cond: (eth_tx_id = eth_txes.id)
   ->  Index Scan using idx_eth_receipts_unique on eth_receipts  (cost=0.28..0.89 rows=1 width=1241) (actual time=0.006..0.006 rows=1 loops=1)
         Index Cond: (tx_hash = eth_tx_attempts.hash)
 Planning Time: 0.804 ms
 Execution Time: 0.081 ms
```
I think it looks pretty good, the `Index Only Scan` on pipeline_runs_suspended is able to eliminate scope (and fail early if there's no suspended runs).

Reply:
This is looking good to me